Generate default constructor for RESTEasy classes if needed

geoand opened the pull request:
This allows users to use constructor injection in their RESTEasy resources.
For a complete background on why this is a proper solution, see:
https://issues.jboss.org/browse/RESTEASY-2183

Fixes: #1314

geoand committed:
Generate default constructor for RESTEasy classes

This allows users to use constructor injection in their code.
For a complete background on this see:
https://issues.jboss.org/browse/RESTEASY-2183

Fixes: #1314

gsmet commented on integration-tests/main/src/main/java/io/quarkus/example/rest/TestResourceWithConstructorInjection.java:
```diff
@@ -0,0 +1,20 @@
+
+    final Someservice service;
+
+    public TestResourceWithConstructorInjection(Someservice service) {
```
@geoand could you add a test with a class having a default constructor and constructor injection too so that we can check it works OK?

geoand replied:
Of course, on it

geoand replied:
Fixed

mkouba commented on extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java:
```diff
@@ -309,13 +316,18 @@ public void build(
                         resources.add(className);
                         reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, className));
+
+                        if (!clazz.hasNoArgsConstructor()) {
```
I'm quite sure we can do this for every resource class. JAX-RS does support some kind of constructor injection, or? Actually, I found a bug in our ArC integration - the following constructor results in unsatisfied dependency:
```java
public TestResourceWithConstructorInjection(@Context Application application) {
   this.application = application;
}
```
Because we automatically consider this a CDI bean constructor.

mkouba replied:
Hm, by the way if I annotate the class with `@ApplicationScoped` I get: 
```
javax.enterprise.inject.spi.DefinitionException: Normal scoped beans must declare a non-private constructor with no parameters
```
Because ArC is analyzing the original class (jandex view) and not the modified one.

mkouba replied:
Finally, the generated no-arg constructor would not be usable - it does not initialize the final fields. So I would expect some ugly exception when it's used...

geoand replied:
I think I am missing some context here. I there a reason one would annotate a `@Path` resource with `@ApplicationScoped`?

The tests currently pass with the `final` fields, there is no problem at all. I haven't tried with `@ApplicationScoped` however.

Also, I didn't really understand the first comment, do you mean "I',m *not* quite sure"? 

So could this all just boil down to the following: Only generate a default constructor when there is no `@ApplicationScoped` annotation (and perhaps a few other constraints)?

mkouba replied:
> I there a reason one would annotate a @Path resource with @ApplicationScoped?

It is allowed. In fact, in CDI world `@ApplicationScoped` is usually used instead of `@Singleton`.

> The tests currently pass...

Yep, I think it passes because the constructor is not used at all.

> Also, I didn't really understand the first comment...

Ah, sorry, yes it should be: "I'm NOT quite sure "

> So could this all just boil down to the following...

I'm not sure. We should think about it...

geoand replied:
My idea is to make the common case easy for developes - meaning that when they write a JAX-RS resource they can use constuctor injection (this is basically what Spring MVC users expect). 
So I would propose that we find the most restricted set of use cases where this is possible ( for example no `@ApplicationScoped` and I don't know what else :) ) and then later on if possible lift these restrictions. WDYT?

kenfinnigan replied:
Is it that bad to require a constructor be added?

Trying to see what the benefits are, and whether they're outweighed by adding "magic" all over the place

geoand replied:
@kenfinnigan this is the first thing that my local JUG leader (and Java Champion Paris Apostolopoulos) complained about for Quarkus. Spring folks are soooooooo used to not needing the extra boilerplate.

kenfinnigan replied:
Ok, but I don't necessarily like the idea of sprinkling magic all over the place just to get Spring folks to look at Quarkus

geoand replied:
It's only magic on the implementation side. For a user perspective having constructor injection work all over the place, makes perfect sense I think. For example, Arc supports constructor injection OOTB, so why not have this support everywhere?

kenfinnigan replied:
As we're trying to remove code as much as possible to reduce size, it's more a concern about this type of thing being added all over the place in Quarkus to support a specific use case, potentially at the detriment of others.

Now that may not be the case here. I'm just cautioning against adding "magic" everywhere without good cause

geoand replied:
:+1: for not adding stuff were we don't need it! 
I am no expert in what overhead some constructors would add, but I would _guess_ that it wouldn't be too much - especially if we only add it under very certain circumstances. In this case I think the added developer joy, would warrant sprinkling a tiny bit of magic.

FroMage replied:
Well, we're adding default constructors to JPA entities in `Panache` and we've already noticed that they're also useful for JSON entities. All this could be done in a new `RESTEasy with Panache` extension, but TBH I don't see how we can justify requiring the user write this constructor when it will be exactly the same as the generated one and it's required and enough to fix these use-cases. That seems to me to be the very definition of "boilerplate".

mkouba replied:
I don't mind adding these constructors but they need to be valid, i.e. if a class declares a final field the constructor must initialize it to a default value as defined by JLS. 

The other problem is with re-indexing the modified class. I don't know how exactly do our bytecode transformers work but I suppose we don't replace the original `ClassInfo` from the index of a pre-indexed dependency? (which is one of the protomoted ways to add a dependency to the application index)

geoand replied:
@mkouba Is it even worth the trouble to generate a proper constructor (instead of just default one) even if the tests never fail? Also, do we need to worry about indexing the class? Are we really gonna need to know about the synthetic no-arg constructor anywere?

mkouba replied:
> Is it even worth the trouble to generate a proper constructor (instead of just default one) even if the tests never fail?

Have you tried to use the generated constructor? If you only want to fool validation logic of some framework the it's ok but for example in case of CDI we do use the no-args constructor to instantiate client proxies. 

> Also, do we need to worry about reindexing.

ArC validation is based on jandex. So if there is a bean class that has normal scope and does not declare a no-args constructor the validation should fail (no matter what the real class look like).

geoand replied:
> > Is it even worth the trouble to generate a proper constructor (instead of just default one) even if the tests never fail?
> 
> Have you tried to use the generated constructor? If you only want to fool validation logic of some framework the it's ok but for example in case of CDI we do use the no-args constructor to instantiate client proxies.
> 

In this case it seems we are simply fooling RESTEasy - see [this](https://issues.jboss.org/browse/RESTEASY-2183?focusedCommentId=13715781). 

The more I think about it the more I think we should only be doing this if the resource is *not* annotated with `@ApplicationScoped` in order to avoid any suprises.
There should probably also be a restriction that there should not be a superclass other than `Object`.  

> > Also, do we need to worry about reindexing.
> 
> ArC validation is based on jandex. So if there is a bean class that has normal scope and does not declare a no-args constructor the validation should fail (no matter what the real class look like).

So if we only do this constructor generation for `Singleton` scope we can get around this restriction as well?

geoand replied:
I pushed a change that now only adds the default constuctor if there is no other annotation other than `@Path` and the resource doesn't extend from any class - This is should be the strictect possible (but most common I would say) use case.
@mkouba Do you think it is safe now?

mkouba replied:
> So if we only do this constructor generation for Singleton scope we can get around this restriction as well?

Yes, `@Singleton` does not require a client proxy.

> Do you think it is safe now?

Well, the constructor would still not be valid if a class contains some final fields but it should be safe enough to fool RESTEasy ;-).

geoand replied:
My thoughts exactly :)

geoand replied:
@mkouba WDYT, should we merge?

mkouba replied:
I'm +0. It looks like a lot of code to support only a very specific simple use case. @gsmet @kenfinnigan WDYT?

geoand replied:
One last thing if I may say so: The use case maybe simple and specific, but it goes a long way to create a consistent experience IMHO. Users should not have to add boilerplate (if that can be avoided) because of limitation of framework - they honestly don't care. Think of it this way: Why should a JAX-RS resource be any different from a user perspective than all the `Singleton` beans he/she adds to their app that work perfectly with constructor injection without requiring a no-arg constructor?

geoand commented on integration-tests/main/src/test/java/io/quarkus/example/test/TestResteasyConstructorsTestCase.java:
```diff
@@ -0,0 +1,22 @@
+import io.restassured.RestAssured;
+
+@QuarkusTest
+public class TestResteasyConstructorsTestCase {
```
Yeah makes sense, I'll do that once we figure out how to address the rest of your comments :)

geoand replied:
Added this test as well

mkouba commented on extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java:
```diff
@@ -344,6 +357,45 @@ public void build(
+                }
+                // again keep it very very simple, we know that there is the @Path annotation, so don't try to generate the
+                // default constructor if there are others (basically we want to avoid surprises with any CDI scopes)
+                if (classInfo.classAnnotations().size() > 1) {
```
Hm, so it will stop working if I add e.g. `@Consumes("application/json")` which is a very common use case.

geoand replied:
Good catch! What other common cases can you think of that we should take into account?

mkouba replied:
Any jax-rs annotation?

geoand replied:
Makes sense!

geoand replied:
Fixed

gsmet commented:
@geoand where are we on this? Should @mkouba review it again?

geoand commented:
@gsmet I have addressed the concerns that @mkouba put forth.

All we need is decision is wether we want this in or not (see https://github.com/quarkusio/quarkus/pull/1875#discussion_r273430725) I strongly believe that it should be (https://github.com/quarkusio/quarkus/pull/1875#discussion_r273814876)
